alacritty/src/renderer/mod.rs

```diff
@@ -82,588 +60,98 @@ impl From<ShaderError> for Error {
-    program: TextShaderProgram,
-    vao: GLuint,
-    ebo: GLuint,
-    vbo_instance: GLuint,
-    atlas: Vec<Atlas>,
-    current_atlas: usize,
-    active_tex: GLuint,
-    batch: Batch,
-
-    rect_renderer: RectRenderer,
-}
-
-#[derive(Debug)]
-pub struct RenderApi<'a> {
-    active_tex: &'a mut GLuint,
-    batch: &'a mut Batch,
-    atlas: &'a mut Vec<Atlas>,
-    current_atlas: &'a mut usize,
-    program: &'a mut TextShaderProgram,
-    config: &'a UiConfig,
-}
-
-#[derive(Debug)]
-pub struct LoaderApi<'a> {
-    active_tex: &'a mut GLuint,
-    atlas: &'a mut Vec<Atlas>,
-    current_atlas: &'a mut usize,
-}
-
-#[derive(Debug, Default)]
-pub struct Batch {
-    tex: GLuint,
-    instances: Vec<InstanceData>,
-}
-
-impl Batch {
-    #[inline]
-    pub fn new() -> Self {
-        Self { tex: 0, instances: Vec::with_capacity(BATCH_MAX) }
-    }
-
-    pub fn add_item(&mut self, cell: &RenderableCell, glyph: &Glyph) {
-        if self.is_empty() {
-            self.tex = glyph.tex_id;
-        }
-
-        let mut cell_flags = RenderingGlyphFlags::empty();
-        cell_flags.set(RenderingGlyphFlags::COLORED, glyph.multicolor);
-        cell_flags.set(RenderingGlyphFlags::WIDE_CHAR, cell.flags.contains(Flags::WIDE_CHAR));
-
-        self.instances.push(InstanceData {
-            col: cell.point.column.0 as u16,
-            row: cell.point.line as u16,
-
-            top: glyph.top,
-            left: glyph.left,
-            width: glyph.width,
-            height: glyph.height,
-
-            uv_bot: glyph.uv_bot,
-            uv_left: glyph.uv_left,
-            uv_width: glyph.uv_width,
-            uv_height: glyph.uv_height,
-
-            r: cell.fg.r,
-            g: cell.fg.g,
-            b: cell.fg.b,
-            cell_flags,
-
-            bg_r: cell.bg.r,
-            bg_g: cell.bg.g,
-            bg_b: cell.bg.b,
-            bg_a: (cell.bg_alpha * 255.0) as u8,
+        point: Point<usize>,
+        fg: Rgb,
+        bg: Rgb,
+        string: &str,
+        size_info: &SizeInfo,
+        glyph_cache: &mut GlyphCache,
+    ) {
+        let cells = string.chars().enumerate().map(|(i, character)| RenderableCell {
+            point: Point::new(point.line, point.column + i),
+            character,
+            zerowidth: None,
+            flags: Flags::empty(),
+            bg_alpha: 1.0,
+            fg,
+            bg,
         });
-    }
-
-    #[inline]
-    pub fn full(&self) -> bool {
-        self.capacity() == self.len()
-    }
-
-    #[inline]
-    pub fn len(&self) -> usize {
-        self.instances.len()
-    }
-
-    #[inline]
-    pub fn capacity(&self) -> usize {
-        BATCH_MAX
-    }
-
-    #[inline]
-    pub fn is_empty(&self) -> bool {
-        self.len() == 0
-    }
 
-    #[inline]
-    pub fn size(&self) -> usize {
-        self.len() * size_of::<InstanceData>()
-    }
-
-    pub fn clear(&mut self) {
-        self.tex = 0;
-        self.instances.clear();
+        match &mut self.text_renderer {
```

Comment:
I'm not sure I understand what do you want here. I need to match enum anyway and call the proper implementation. I can't reuse with_api, since Api are different and you can't make closures generic over argument, only functions. If I implement `draw_cells` on enum it'll still require me to do `Self::Glsl3(renderer) => renderer.draw_cells()`, no?

Reply:
Correct, but you'd only have to do it once. You're currently duplicating the logic because you're calling `draw_calls` in two different locations.